### ISSUE TITLE
indentation of methods/properties on newline

### ISSUE THREAD

**Issue by marneborn** [mooz/js2-mode#162]
There are a couple of issues around methods/properties on new lines.

Sorry for putting several issues in one ticket, but I wanted to make sure that you had the context for them.

I've (mostly) got them "fixed" so that I'm happy, and I'm looking at cleaning them up for general use and wanted to get your input before going to further down that path.

First, a basic issue/question, when there is an assignment where an objects method/property is on a new line, js2-multiline-decl-indentation kicks in. So this gets indented `(+ declaration-indent js2-basic-offset)`. Is this intended?

``` javascript
var iable = isSet
        .toSomething()
        .andSomethingIsDone();
```

Or should it be js2-basic-offset. Ie:

``` javascript
var iable = isSet
    .toSomething()
    .andSomethingIsDone();
```

I prefer the second, but don't care much.

Second, I can't put in extra indents for when the context changes.
From: http://contribute.jquery.org/style-guide/js/#chained-method-calls

> **Chained Method Calls**
> 
> When a chain of method calls is too long to fit on one line, there must be one call per line, with the first call on a separate line from the object the methods are called on. If the method changes the context, an extra level of indentation must be used.
> 
> ``` javascript
> elements
>    .addClass( "foo" )
>    .children()
>        .html( "hello" )
>    .end()
>    .appendTo( "body" );
> ```

What I'm working on now is having tab toggle between any of the legal indentations.

The third issue is that, despite the jQuery guideline, I want the newlined methods/properties to line up with the initial object. This is because when I have promises inside of promise callbacks that extra level indentation is too much (for me). This would be variable controllable.

Instead of this:

``` javascript
returnsAPromise()
    .then(
        function ( val ) {
            alsoReturnsAPromise()
                .then (
                    function () { doSomethingUseful(); },
                    function () { handleRejection (); }
                );
        },
        function ( reason ) {
            handleRejection();
        }
    );
```

I prefer this:

``` javascript
returnsAPromise()
.then(
    function ( val ) {
        alsoReturnsAPromise()
        .then(
            function () { doSomethingUseful(); },
            function () { handleRejection (); }
        );
    },
    function ( reason ) {
        handleRejection();
    }
);
```

Are any of these issues things that you want worked on?
In summary:
1. How do you want the first method/property on a new line to be indented with an assignment?
2. Do you want to be able to use 'tab' to toggle between valid indent levels for chained calls?
3. Do you want a mode that sets the initial indent level for methods/properties on a newline?

I want 2 and 3 and will have that in my clone.
I need the answer to 1 before trying to code them for general consumption.

thanks,
-mikael

**Comment by iezer** [comment]
👍 I'm also interested in 3., did you find a solution?

**Comment by katspaugh** [comment]
My solution was to tell `js-mode` to not indent multi-line declarations in any special way:

```
    (advice-add 'js--multi-line-declaration-indentation :around (lambda (orig-fun &rest args) nil))
```

**Comment by iezer** [comment]
I found adding this to my config fixed the issue:
`'(js-expr-indent-offset -4)`

You can also customize that property by doing `M-x customize-group RET js RET` and changing `Hide Js Expr Indent Offset`. The value should be negative however many spaces you use for tabs, in your example `4`.
